### PR TITLE
[screen-orientation][ios] Screen Orientation module audit

### PR DESCRIPTION
**Pull request by behenate**
# Why

Screen Orientation module is written in old modules API.
Screen orientation doesn't work correctly on iOS 16+  https://github.com/expo/expo/issues/21974, https://github.com/expo/expo/issues/21044
Various minor bugs (see below).

ENG-8363

# How

Migration: Followed the usual migration steps.

Bugfixes: 
- Moved to new screen orientation API on iOS 16
- Most of the issues below were fixed by using `supportedInterfaceOrientationsFor` in AppDelegates
- Fixed the bug on iOS <= 15 where safe area view would fail when locked in orientaion X, device in orientation Y and pulled down the quick settings menu
- Now when locked to portrait/landscape the status and navigation bars won't rotate freely.
- In general the orientation behaviour is now basically the same as in native apps.
- Due to recent changes dev-menu causes the app to crash on assert in `requestOverlayMetricsIfNeeded` function. When in landscape mode the notch height would be returned as negative, which causes the assertion to fail. I corrected it to always return values greater or equal than zero.

### Here are some videos comparing old and new behaviours:
Old version is on the left, post-audit version is on the right. If the videos aren't loading try using Google Chrome.
| | |
| ------------- | ------------- |
| iOS <=15 | iOS <= 15  |
| <video src="https://user-images.githubusercontent.com/31368152/234654161-725fb09a-9539-4d9a-ae34-5af0287f5b5d.mov">  | <video src="https://user-images.githubusercontent.com/31368152/234654551-341af83d-ca1c-4a9d-877a-c0e869218265.mov">|
| iOS <=15 | iOS >= 16  |
|<video src="https://user-images.githubusercontent.com/31368152/234654565-c47919fe-4f62-4f2b-a95a-3726420927f5.mov"/> | <video src="https://user-images.githubusercontent.com/31368152/234659468-2399e5bb-7fab-4ab9-96db-6c221954ef79.mov" />  |

# Test Plan

Tested on simulator and physical devices (iOS 15 and 16).

### CODE REVIEW

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      if !allowedOrientationsMask.isSupportedByDevice() {
+        promise.reject(UnsupportedOrientationLockException(nil))
+        return
+      }
+
+      screenOrientationRegistry.setMask(allowedOrientationsMask, forModule: self)
+      promise.resolve()
+    }
+
+    AsyncFunction("getOrientationLockAsync") { (promise: Promise) in
```
Comment (lukmccall) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:51]: and here

**Review by lukmccall — CHANGES_REQUESTED** [review]

File: packages/expo-modules-core/CHANGELOG.md [hunk 2]
```diff
@@ -24,6 +24,7 @@
 - [iOS] Use `jsi::WeakObject` for weak objects on Hermes. ([#21986](https://github.com/expo/expo/pull/21986) by [@tsapeta](https://github.com/tsapeta))
 - [iOS] Removed legacyViewManager references from ExpoFabricView. ([#22089](https://github.com/expo/expo/pull/22089) by [@gabrieldonadel](https://github.com/gabrieldonadel))
 - [iOS] Dropped `proxiedProperties` prop. ([#22280](https://github.com/expo/expo/pull/22280) by [@tsapeta](https://github.com/tsapeta))
+- [iOS] Added Interface Geometry Management section to ExpoAppDelegate. ([#22152](https://github.com/expo/expo/pull/22152) by [@behenate](https://github.com/behenate))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:27]: It would be great if you could make a separate PR for that feature 😉

File: packages/expo-dev-menu/CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,7 @@
 - Fixed reload crash when expo-dev-menu is turned off. ([#21279](https://github.com/expo/expo/pull/21279) by [@jayshah123](https://github.com/jayshah123))
 - Fix JS entry file in development builds. ([#21984](https://github.com/expo/expo/pull/21984) by [@gabrieldonadel](https://github.com/gabrieldonadel))
 - Update the start script to dynamically locate the packager IP from any network interface. ([#21977](https://github.com/expo/expo/pull/21977) by [@gabrieldonadel](https://github.com/gabrieldonadel))
+- [iOS] Fix assert in `requestOverlayMetricsIfNeeded` failing when the device is rotated to landscape orientation. ([#22152](https://github.com/expo/expo/pull/22152) by [@behenate](https://github.com/behenate))
```
Comment (tsapeta) [inline packages/expo-dev-menu/CHANGELOG.md:16]: Same here – would be great to make a separate PR for changes in expo-dev-menu

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+      if !orientationMask.isSupportedByDevice() {
+        throw UnsupportedOrientationLockException("\(orientationLock.rawValue)")
+      }
+      screenOrientationRegistry.setMask(orientationMask, forModule: self)
+    }
+
+    AsyncFunction("lockPlatformAsync") { (allowedOrientations: [ModuleOrientation]) in
+      var allowedOrientationsMask: UIInterfaceOrientationMask = []
+      for allowedOrientation in allowedOrientations {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:28]:
```suggestion

      for allowedOrientation in allowedOrientations {
```

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        let orientation = allowedOrientation.toInterfaceOrientation()
+        let orientationMask = orientation.toInterfaceOrientationMask()
+        if orientationMask.isEmpty {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:31]:
```suggestion

        if orientationMask.isEmpty {
```

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+          throw InvalidOrientationLockException()
+          return
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:33]: No need to return if you already thrown an exception

```suggestion
```

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        allowedOrientationsMask.insert(orientationMask)
+      }
+
+      if !allowedOrientationsMask.isSupportedByDevice() {
+        throw UnsupportedOrientationLockException(nil)
+        return
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:41]:
```suggestion
```

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        let orientationMask = orientation.toInterfaceOrientationMask()
+        if orientationMask.isEmpty {
+          throw InvalidOrientationLockException()
+          return
+        }
+
+        allowedOrientationsMask.insert(orientationMask)
+      }
+
+      if !allowedOrientationsMask.isSupportedByDevice() {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:39]: In most cases when you want to exit early from the function when certain conditions are not met, it's better to use `guard..else`
```suggestion
      guard allowedOrientationsMask.isSupportedByDevice() else {
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 9]
```diff
@@ -0,0 +1,239 @@
+import Foundation
+import ExpoModulesCore
+protocol OrientationListener {
+  func screenOrientationDidChange(_ orientation: UIInterfaceOrientation)
+}
+
+// This singleton holds information about desired orientation for every app which uses expo-screen-orientation.
+// Marked @objc and public, because this it is also used in EXAppViewController.
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:8]:
```suggestion
/**
 The singleton that holds information about desired orientation for every app which uses expo-screen-orientation.
 Marked @objc and public, because this it is also used in EXAppViewController.
 */
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    NotificationCenter.default.addObserver(
+      self,
+      selector: #selector(self.handleDeviceOrientationChange(notification:)),
+      name: UIDevice.orientationDidChangeNotification,
+      object: UIDevice.current
+    )
+
+    DispatchQueue.main.async {
+      UIDevice.current.beginGeneratingDeviceOrientationNotifications()
+    }
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:48]: Do we need to switch the thread and run it asynchronously? Isn't the initializer already called on the main thread?

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+  }
+
+  deinit {
+    EXUtilities.performSynchronously {
+      UIDevice.current.endGeneratingDeviceOrientationNotifications()
+    }
+  }
+
+  // MARK: affecting screen orientation
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:65]: `-` makes the difference in Xcode's method navigator. We used to use the line separator.
```suggestion
  // MARK: - Affecting screen orientation
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+  // Rotates the view to currentScreenOrientation (or default orientation from the orientationMask)
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:66]:
```suggestion
  /**
   Rotates the view to currentScreenOrientation (or default orientation from the orientationMask)
   */
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+  func enforceDesiredDeviceOrientation(withOrientationMask orientationMask: UIInterfaceOrientationMask) {
+    var newOrientation = orientationMask.defaultOrientation()
+
+    if orientationMask.contains(currentScreenOrientation) {
+      newOrientation = currentScreenOrientation
+    }
+
+    if newOrientation != .unknown {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:74]: I would use a guard here to keep indentation level lower

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+      DispatchQueue.main.async { [weak self] in
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:75]: Is it necessary to switch the thread and run it asynchronously?

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+    }
+  }
+
+  func setMask(_ mask: UIInterfaceOrientationMask, forModule module: ScreenOrientationModule) {
+    moduleInterfaceMasks[module] = mask
+    enforceDesiredDeviceOrientation(withOrientationMask: mask)
+  }
+
+  // MARK: getters
+  // Gets the orientationMask for the current module. Also used for Expo Go in EXAppViewController
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:104]:
```suggestion
  /**
   Gets the orientationMask for the current module. Also used for Expo Go in `EXAppViewController`.
   */
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+    for moduleMask in moduleInterfaceMasks {
+      mask = mask.intersection(moduleMask.value)
+    }
+
+    return mask
+  }
+
+  // MARK: events
+
+  // Called when iOS sends and OrientationDidChangeNotification
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:123]:
```suggestion
  /**
   Called when iOS sends and OrientationDidChangeNotification.
   */
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+  @objc func handleDeviceOrientationChange(notification: Notification) {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:124]:
```suggestion
  @objc
  func handleDeviceOrientationChange(notification: Notification) {
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+      let portraitMask = currentOrientationMask.intersection([.portrait, .portraitUpsideDown])
+
+      // Mask allows only proper portrait - we know that the device is in either proper portrait or upside down
+      // we deduce it is proper portrait.
+      if portraitMask == .portrait {
+        newScreenOrientation = .portrait
+      }
+      // Mask allows only upside down portrait - we know that the device is in either proper portrait or upside down
+      // we deduce it is upside down portrait.
+      else if portraitMask == .portraitUpsideDown {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:179]: I think in that case it's better to put the comment inside the `if` block so we don't change/mix our formatting conventions.
```suggestion
      } else if portraitMask == .portraitUpsideDown {
        // Mask allows only upside down portrait - we know that the device is in either proper portrait or upside down
        // we deduce it is upside down portrait.
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+        newScreenOrientation = .portraitUpsideDown
+      }
+      // Mask allows portrait or upside down portrait - we can try to deduce orientation
+      // from device orientation.
+      else if currentDeviceOrientation == .portrait || currentDeviceOrientation == .portraitUpsideDown {
+        newScreenOrientation = currentDeviceOrientation
+      }
+    } else if (verticalSizeClass == .compact && horizontalSizeClass == .compact)
+      || (verticalSizeClass == .regular && horizontalSizeClass == .regular)
+      || (verticalSizeClass == .compact && horizontalSizeClass == .regular) {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:189]: For more complex conditions like this, I would recommend extracting them to a separate function

File: packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift [hunk 20]
```diff
@@ -0,0 +1,114 @@
+import ExpoModulesCore
+
+// Modules are not hashable. We are using a dictionary (moduleInterfaceMasks), where modules are keys.
+extension ScreenOrientationModule: Hashable {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift:4]: Let's move this to `ScreenOrientationModule.swift`

File: packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+// https://medium.com/@cafielo/how-to-detect-notch-screen-in-swift-56271827625d
+internal var doesDeviceHaveNotch = {
+  var bottomSafeAreaInsetIsPositive = false
+  EXUtilities.performSynchronously {
+    bottomSafeAreaInsetIsPositive = (UIApplication.shared.delegate?.window??.safeAreaInsets.bottom ?? 0.0) > 0.0
+  }
+  return bottomSafeAreaInsetIsPositive
+}()
+
+internal func isIPad() -> Bool {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift:27]: The identical `isPad` function is still there 😉

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+        let supportedOrientationMask = orientationMask.intersection(wrappedSingleOrientation)
+        if !supportedOrientationMask.isEmpty {
+          let supportedOrientation = supportedOrientationMask.toUIInterfaceOrientation()
+          allowedOrientations.append(ModuleOrientation.from(orientation: supportedOrientation).rawValue)
+        }
+      }
+      return allowedOrientations
+    }
+
+    AsyncFunction("supportsOrientationLockAsync") {(orientationLock: ModuleOrientationLock) -> Bool in
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:67]:
```suggestion
    AsyncFunction("supportsOrientationLockAsync") { (orientationLock: ModuleOrientationLock) -> Bool in
```

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+
+    Events("expoDidUpdateDimensions")
+
+    AsyncFunction("lockAsync") { (orientationLock: ModuleOrientationLock) in
+      let orientationMask = orientationLock.toInterfaceOrientationMask()
+      guard !orientationMask.isEmpty else {
+        throw InvalidOrientationLockException()
+      }
+
+      if !orientationMask.isSupportedByDevice() {
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:20]:
```suggestion
      guard orientationMask.isSupportedByDevice() else {
```

File: packages/expo-screen-orientation/ios/ScreenOrientationViewController.swift [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+
+  convenience init(defaultScreenOrientationFromPlist: Void) {
+    guard let orientationString = Bundle.main.object(forInfoDictionaryKey: defaultScreenOrientationMask) as? String else {
+      self.init(defaultOrientationMask: .portrait)
+      return
+    }
+
+    // TODO: When printing to errors to JS available print the warnings there (@behenate)
+    guard let mask = plistStringToInterfaceOrientationMask(orientationString) else {
+      print("Orientation lock string '\(orientationString)' provided in Info.plist does not correspond to a valid orientation mask. Application will default to portrait orientation lock.")
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationViewController.swift:22]:
```suggestion
      log.warn("Orientation lock string '\(orientationString)' provided in Info.plist does not correspond to a valid orientation mask. Application will default to portrait orientation lock.")
```

File: packages/expo-screen-orientation/ios/ScreenOrientationViewController.swift [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+      self.init(defaultOrientationMask: .portrait)
+      return
+    }
+
+    guard mask.isSupportedByDevice() else {
+      print("Orientation lock string '\(orientationString)' provided in Info.plist is not supported by the device. Application will default to portrait orientation lock.")
```
Comment (tsapeta) [inline packages/expo-screen-orientation/ios/ScreenOrientationViewController.swift:28]:
```suggestion
      log.warn("Orientation lock string '\(orientationString)' provided in Info.plist is not supported by the device. Application will default to portrait orientation lock.")
```

**Review by tsapeta — CHANGES_REQUESTED** [review]

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift (hunk 14 above)
Comment (behenate) [reply packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:75]: The functions inside have to be run on the main thread

**Comment by Wilson-Lim** [comment]
Is there a way to use this fix without waiting for Pull Request being merged ?

File: ios/Exponent/Kernel/Views/EXAppViewController.m [hunk 26]
```diff
@@ -578,12 +578,10 @@ - (UIInterfaceOrientationMask)supportedInterfaceOrientations
     return [super supportedInterfaceOrientations];
   }
 
-#if __has_include(<EXScreenOrientation/EXScreenOrientationRegistry.h>)
-  EXScreenOrientationRegistry *screenOrientationRegistry = (EXScreenOrientationRegistry *)[EXModuleRegistryProvider getSingletonModuleForClass:[EXScreenOrientationRegistry class]];
-  if (screenOrientationRegistry && [screenOrientationRegistry requiredOrientationMask] > 0) {
-    return [screenOrientationRegistry requiredOrientationMask];
+  if ([ScreenOrientationRegistry.shared requiredOrientationMask] > 0){
```
Comment (lukmccall) [inline ios/Exponent/Kernel/Views/EXAppViewController.m:581]:
```suggestion
  if ([ScreenOrientationRegistry.shared requiredOrientationMask] > 0) {
```

File: packages/expo-screen-orientation/ios/ScreenOrientationModule.swift [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+    AsyncFunction("lockPlatformAsync") { (allowedOrientations: [ModuleOrientation]) in
+      var allowedOrientationsMask: UIInterfaceOrientationMask = []
+
+      for allowedOrientation in allowedOrientations {
+        let orientation = allowedOrientation.toInterfaceOrientation()
+        let orientationMask = orientation.toInterfaceOrientationMask()
+
+        guard !orientationMask.isEmpty else {
+          throw InvalidOrientationLockException()
+      }
```
Comment (lukmccall) [inline packages/expo-screen-orientation/ios/ScreenOrientationModule.swift:37]:
```suggestion
        }
```

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 28]
```diff
@@ -0,0 +1,248 @@
...
+
+    if orientationMask.contains(currentScreenOrientation) {
+      newOrientation = currentScreenOrientation
+    }
+
+    guard newOrientation != .unknown else {
+      return
+    }
+
+    RCTExecuteOnMainQueue { [weak self] in
```
Comment (lukmccall) [inline packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:85]:
```suggestion
    DispatchQueue.main.async { [weak self] in
```
I would use the `DispatchQueue.main` instead of `RCTExecuteOnMainQueue`. The effect is the same, but we're using fewer things from the react native, which is good.

File: packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+
+internal func isPad() -> Bool {
+  return UIDevice.current.userInterfaceIdiom == .pad
+}
+
+// https://medium.com/@cafielo/how-to-detect-notch-screen-in-swift-56271827625d
+internal var doesDeviceHaveNotch = {
+  var bottomSafeAreaInsetIsPositive = false
+  EXUtilities.performSynchronously {
+    bottomSafeAreaInsetIsPositive = (UIApplication.shared.delegate?.window??.safeAreaInsets.bottom ?? 0.0) > 0.0
```
Comment (lukmccall) [inline packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift:11]:
```suggestion
    bottomSafeAreaInsetIsPositive = (UIApplication.shared.delegate?.window?.safeAreaInsets.bottom ?? 0.0) > 0.0
```
I didn't know that you can chain the `?` operator 😅

File: packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+    case .landscapeLeft: return .landscapeLeft
+    case .landscapeRight: return .landscapeRight
+    default: return .unknown
+    }
+  }
+
+  internal func contains(_ orientation: UIInterfaceOrientation) -> Bool {
+    // This is how the mask is created from the orientation
+    let maskFromOrientation = orientation.toInterfaceOrientationMask()
+    return self.contains(maskFromOrientation)
```
Comment (lukmccall) [inline packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift:30]:
```suggestion
    return contains(maskFromOrientation)
```
I have mixed feelings about it. Using the `self` keyword may not be essential, but it could enhance the readability of the code. It's ultimately your decision, but if you choose to delete it, please do so for all instances in this file.

**Review by lukmccall — APPROVED** [review]

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift (hunk 28 above)
Comment (tsapeta) [reply packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:85]: That was me who recommended to use `RCTExecuteOnMainQueue` here. There is a small difference – if it's currently on the main thread, the work will be executed synchronously without deferring it. I'm fine with using it from RN for now, but we should make our own utils for that.

File: packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift (hunk 30 above)
Comment (behenate) [reply packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift:30]: I think I prefer the version with self because this is an extension and it might not be obvious if we are referencing external function or a method right away (unless someone knows the syntax highlighting colors of xcode by heart 🤓). @tsapeta which version do you prefer?

File: packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift (hunk 29 above)
Comment (behenate) [reply packages/expo-screen-orientation/ios/ScreenOrientationUtilities.swift:11]: The delegate returns a `UIWindow??` type when referencing `delegate.window` so the chaining of `?` operator is necessary. It's unclear to me why this is the case though.

**Review by tsapeta — APPROVED** [review]

File: packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift [hunk 31]
```diff
@@ -0,0 +1,248 @@
...
+      } else if currentDeviceOrientation == .portrait || currentDeviceOrientation == .portraitUpsideDown {
+        // Mask allows portrait or upside down portrait - we can try to deduce orientation
+        // from device orientation.
+        newScreenOrientation = currentDeviceOrientation
+      }
+    } else if traitCollection.isLandscape() {
+      // From trait collection, we know that screen is in landscape left or right orientation.
+      let landscapeMask = currentOrientationMask.intersection(.landscape)
+
+      if landscapeMask == .landscapeLeft {
```
[the remark this answers is not in the document]
Comment (behenate) [reply packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift:203]: Adding a `switch` requires adding an `if` statement inside of the `default` block. I wrote the `switch` version and I think the current version looks a bit cleaner.

**Review by tsapeta — APPROVED** [review]

**Comment by watadarkstar** [comment]
@tsapeta @behenate Can I try this out in a nightly build? Does Expo build nightly?

```
yarn add expo-screen-orientation@nightly
```

**Comment by tsapeta** [comment]
@watadarkstar we don't build nightlies yet, but I'm going to publish some packages today, including `expo-screen-orientation` 😉 I'll let you know here!

**Comment by tsapeta** [comment]
@watadarkstar
It's out in `6.0.0-beta.1`, you can use `yarn add expo-screen-orientation@next` to install it 🙂
Let us know your feedback 😊

**Comment by liquidvisual** [comment]
I couldn't seem to get this working when installing the pods, with SDK44 - can't update at the moment unfortunately:

```
Command `pod install` failed.
└─ Cause: Failed to load 'ExpoScreenOrientation' podspec: 
[!] Invalid `ExpoScreenOrientation.podspec` file: undefined method `install_modules_dependencies' for Pod:Module.
```

**Comment by tsapeta** [comment]
@liquidvisual
I'm sure it will not work with that old SDK version. The `install_modules_dependencies` method comes from React Native and was added probably in 0.71. I would recommend upgrading to this version and Expo SDK 48 (49 will be out in a few weeks though).

**Comment by singh-sukhmanjit** [comment]
Am able to rotate orientation using `ScreenOrientation.lockAsync(orientationLock)`, but it doesnt lock it. If I rotate my phone then orientation changes accordingly.

**Comment by behenate** [comment]
@singh-sukhmanjit What iOS version are you using, also are you using expo go or bare workflow? Code which you are using to lock the screen would be helpful too.

**Comment by watadarkstar** [comment]
@behenate @tsapeta I tested this on an iPhone 12 mini on iOS 16.2 using the package 6.0.0-beta.1 and its working great. Thank you! 🙏

**Comment by singh-sukhmanjit** [comment]
@behenate Am trying it on iPad (8th generation) iPadOS 16.1 with Expo SDK 48 bare workflow. Am using `ScreenOrientation.lockAsync(ScreenOrientation.OrientationLock.LANDSCAPE_LEFT);` to lock the screen. Info.plist includes:
```
<key>UISupportedInterfaceOrientations</key>
<array>
    <string>UIInterfaceOrientationPortrait</string>
    <string>UIInterfaceOrientationLandscapeLeft</string>
    <string>UIInterfaceOrientationLandscapeRight</string>
</array>
```

**Comment by behenate** [comment]
@singh-sukhmanjit Have you set the "requireFullScreen" option to true in [app.json](https://docs.expo.dev/versions/latest/sdk/screen-orientation/?utm_source=google&utm_medium=cpc&utm_content=performancemax&gclid=Cj0KCQjw7aqkBhDPARIsAKGa0oIz3ZBEmvvDk0BsILPF7zgghdE6SvNsc09dZ3hGYHwepvJgChKEujkaAnNSEALw_wcB#example-appjson-with-config-plugin)?
You can also try setting it to true directly in Info.plist.

**Comment by singh-sukhmanjit** [comment]
@behenate yes, it is set but still the behavior is same.

**Comment by jasonz1987** [comment]
still not work no ios 16.2 iphone simulator  (bare workflow)

xcode log

> 2023-06-17 12:16:14.903751+0800 app[32163:174205] [Orientation] BUG IN CLIENT OF UIKIT: Setting UIDevice.orientation is not supported. Please use UIWindowScene.requestGeometryUpdate(_:)


my env：

```
system:
    OS: macOS 12.6.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 27.57 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.10.0 - ~/.nvm/versions/node/v16.10.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.10.0/bin/yarn
    npm: 7.24.0 - ~/.nvm/versions/node/v16.10.0/bin/npm
    Watchman: 2023.05.22.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.12.1 - /Users/jasonz/.rbenv/shims/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    Android SDK:
      Android NDK: 23.0.7344513-beta4
  IDEs:
    Android Studio: 2021.2 AI-212.5712.43.2112.8609683
    Xcode: 14.2/14C18 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.18 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.2.0 => 18.2.0
    react-native: 0.71.8 => 0.71.8
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found
```


package.json

```
   "expo-screen-orientation": "^6.0.0-beta.1",
   "expo": "~48.0.18",
```

**Comment by behenate** [comment]
@jasonz1987
I managed to replicate your issue. The problem is that the `expo-modules-core` part of the audit hasn't been published yet and in some cases its crucial for the `screen-orientation` module to work correctly. The necessary code change is small and has already been merged, but you will have to wait a bit until we publish it https://github.com/expo/expo/pull/22599. 
For now you can try to paste the code from the PR into `ExpoAppDelegate.swift` rebuild the project and see if it works correctly for you. 

It is also possible that your `app.json` is not configured correctly.  If your `app.json` contains the following line: `    "orientation": "portrait"` you should remove it and then add configuration line for the config plugin with your desired initial orientation lock [example](https://docs.expo.dev/versions/latest/sdk/screen-orientation/#example-appjson-with-config-plugin).  Then run `npx expo prebuild` and rebuild your project.


As for the error it is most likely caused by `react-native-screens` (in your project it is most likely used by `react-navigation`) which is still trying to use the old orientation API on newer iOS versions, this shouldn't have any impact on `expo-screen-orientation`. If you install `3.21.0` version of `react-native-screens` the error shouldn't appear anymore.

**Comment by jasonz1987** [comment]
@behenate the pr #22599 works, thanks a a lot.

**Comment by ycldz** [comment]
Expo: "~48.0.15"
expo-screen-orientation: "~5.1.1"
IOS Simulator: Iphone 14 Pro Max / v16.4
Real Device: Iphone 11 / v16.6

app.json in orientation deleted

app.json plugins in  added
[
   "expo-screen-orientation",
   {
        "initialOrientation": "DEFAULT"
   }
],

await ScreenOrientation.lockAsync(ScreenOrientation.OrientationLock.LANDSCAPE);

It works fine on the simulator but not on the real device, has anyone solved this problem?

**Comment by behenate** [comment]
@ycldz 
Hi, please upgrade to SDK 49 and screen orientation 6.0.5 and try again. (Don't upgrade only ScreenOrientation as it relies on some changes made in SDK 49)

**Comment by ycldz** [comment]
@behenate unfortunately it doesn't work on expo 49 either.

**Comment by ycldz** [comment]
@behenate I save the code in the editor, the screen does not refresh on the device and the console.log code does not work thank you :D

**Comment by behenate** [comment]
@ycldz Are you using managed or bare workflow? I tested in a bare workflow with this app https://github.com/behenate/screen-orientation-demo and it seems to be working fine

**Comment by Darkhorse-Fraternity** [comment]
image
it doesn't work on expo 49 either.

**Comment by Darkhorse-Fraternity** [comment]
> @ycldz Are you using managed or bare workflow? I tested in a bare workflow with this app https://github.com/behenate/screen-orientation-demo and it seems to be working fine

The reproduction of this issue requires the simultaneous use of  screen-orientation and react-navigation.

**Comment by Darkhorse-Fraternity** [comment]
```
export const useLandscape = () => {
  useEffect(() => {
      void InteractionManager.runAfterInteractions(() => {
        void ScreenOrientation.lockAsync(
          ScreenOrientation.OrientationLock.LANDSCAPE_RIGHT,
        );
        // ...long-running synchronous task...
      });

    return () => {
      void InteractionManager.runAfterInteractions(() => {
        void ScreenOrientation.lockAsync(
          ScreenOrientation.OrientationLock.PORTRAIT_UP,
        );
        // ...long-running synchronous task...
      });
    };
  }, []);
};
```
Putting this feature on the page after the push will have a certain probability of recurring.
